### ISSUE TITLE
Considier simplifying modifying dev container definitions by including a shell script

### ISSUE THREAD

**Issue by Chuxel** [microsoft/vscode-dev-containers#916]
Currently the contents of this repository include Dockerfiles that use separate run statements to provide good layering.  However, the syntax is not immediately obvious to those that do not know Docker or, even worse, are new to Linux.

Instead, we could simplify the Docker files down to basically a set of environment variables and arguments and a single script that is executed.  We then include a bash script that users can contribute their own. 

A working example can be found here: https://github.com/microsoft/vscode-dev-containers/tree/clantz/explore-simplification/containers/dotnet/.devcontainer

Any ARGs (or ENVs) you place in the Dockerfile will end up as environment variables in scripts in the `library-scripts` folder. The `install` script knows how to execute scripts in the `options` folder if certain arguments are set and the `base` scripts are only used when building the image we publish (using base.Dockerfile).  Anything else in the librrary-scripts folder ending in `.sh` is executed... so developers can put anything they want in that location.

//cc @2percentsilk @bamurtaugh @chrmarti for thoughts - This is related to the idea of simplifying the use of script-library scripts since similar changes are needed there, but still separate.

**Comment by Chuxel** [comment]
This will require some subtle tweaks to the build system to make work, but they're not bad.

**Comment by joshspicer** [comment]
I think overall this seems like a great idea in terms of making composability easy. 

However as a counter-argument, I could see this being more confusing to people who _are_ well-versed in docker syntax, as it then requires someone more experienced to take the time to learn how to use these "lego blocks"

**Comment by Chuxel** [comment]
Yeah I think we'd include comments just saying that there's a script in a sub-folder you can use to modify the script to help alleviate that part.  Really from a user perspective, they can still modify the Dockerfile as they would normally if that's what they prefer. This just makes it easier for those that don't know the file to do what they need to do.

**Comment by bamurtaugh** [comment]
When comparing https://github.com/microsoft/vscode-dev-containers/blob/clantz/explore-simplification/containers/dotnet/.devcontainer/Dockerfile with https://github.com/microsoft/vscode-dev-containers/blob/main/containers/dotnet/.devcontainer/Dockerfile, I think the working example @chuxel shares does look quite a bit cleaner. 

I think with the right docs + guiding comments, this could really make Dockerfiles and the process of customizing containers feel a lot less intimidating for folks - if I can tell what's in the Dockerfile just by the easy to understand names of environment variables, that feels quite approachable. 

As a less experienced user, I think the most intimidating / uncertain part would be [line 13](https://github.com/microsoft/vscode-dev-containers/blob/clantz/explore-simplification/containers/dotnet/.devcontainer/Dockerfile#L13):

> You can add your own install steps to "library-scripts/user-install-steps.sh" or another file in this folder.

I wouldn't necessarily know what "install steps" are or how to create a script. What am I installing? What steps are required?

I see there's quite comprehensive information about scripts in this repo (i.e. here: https://github.com/microsoft/vscode-dev-containers/tree/main/script-library and within each individual script's readme). Perhaps on https://code.visualstudio.com/docs we could also provide some guidance on scripts in general, i.e. the structure of the library-scripts folder, its purpose, how to create scripts. If we are transitioning to a model extracting more from the Dockerfile and into scripts, it feels like documenting this clearly would be useful.

**Comment by Chuxel** [comment]
> > You can add your own install steps to "library-scripts/user-install-steps.sh" or another file in this folder.
> I wouldn't necessarily know what "install steps" are or how to create a script. What am I installing? What steps are required?

@bamurtaugh Yeah, there's the lego block idea we're hoping to converge on that will actually nail selection of the existing scripts.  

What this file does is give you a bash script for you to put contents in.  So, if you see instructions that look like this for the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-linux.html)...
```
curl "https://s3.amazonaws.com/aws-cli/awscli-bundle.zip" -o "awscli-bundle.zip"
unzip awscli-bundle.zip
sudo ./awscli-bundle/install -i /usr/local/aws -b /usr/local/bin/aws
```
...you can just paste it into the file.  No need to deal with `RUN` and `&&` statements.  There's one sitting there for your own user contents that is run.

**Comment by bamurtaugh** [comment]
Got it! So you just need to know the scripts you want (i.e. steps for installing AWS CLI you show above) and don't need to know the precise Dockerfile syntax. 

I can see in https://github.com/microsoft/vscode-dev-containers/blob/clantz/explore-simplification/containers/dotnet/.devcontainer/library-scripts/user-install-steps.sh where the user would be able to add steps like that.

> But maybe the comment isn't pointing that out well enough... the question is whether this is less confusing to do or not.

I'm trying to think how we could convey that. Maybe just documenting this type of example well in the readme and in docs, and pointing users to docs as well (from the readme and even from the code itself) would help.
